Initial tests of parallel execution, shift changing, etc.

CharliePoole opened the pull request:
This PR is intended to collect fixes for a number of parallelism issues, particularly around use of SetUpFixtures but also with the handling of parallel and non-parallel TestFixtures and the smooth changing of shifts within the dispatcher.

The initial commit sets up some infrastructure and adds a few unit tests of the parallel dispatcher. Those tests are known to fail intermittently, just as user runs are failing intermittently in some cases.

The issues I hope to address in this PR include #1239, ~~#1905~~, #2364, #2388, #2467 and #2471. I'll add "Fixes" notations when I think any of them is fixed.

I'd like it if @nunit/framework-team members followed along and commented.

Fixes #1239 (needs review)
Fixes #2364
Fixes #2395
Fixes #2467
Fixes #2471
Fixes #2478

CharliePoole commented:
The second commit removes saving and restoring of queues entirely from the dispatcher. This feature was added so that non-parallel fixtures containing parallel test cases could run in absolute isolation from other fixtures. The non-parallel fixture gets a complete set of queues all to itself.

I'm removing it so that it can be re-implemented in a cleaner way. It's possible to do this temporarily because our own tests don't present the situation for which the feature was designed - i.e. having a non-parallel fixture with parallel child tests.

CharliePoole commented:
Fixes #2478 - a new issue found while debugging parallel SetUpFixtures

jnm2 commented:
Spotted a typo in the first commit message.

jnm2 commented on src/NUnitFramework/TestBuilder.cs:
```diff
@@ -60,51 +64,19 @@ public static TestSuite MakeFixture(object fixture)
             return suite;
         }
 
-        public static TestSuite MakeParameterizedMethodSuite(object fixture, string methodName)
```
The hard part with reviewing commits this big is that it seems like this refactor will be a very cross-cutting change, but I'm all for streamlining this API. 😃

jnm2 commented on src/NUnitFramework/TestBuilder.cs:
```diff
@@ -161,34 +133,45 @@ public static ITestResult RunTestCase(object fixture, string methodName)
-        // tests that need to access the WorkItem before or after execution.
+        public static CompositeWorkItem CreateWorkItem(Type type)
+        {
+            return CreateWorkItem(MakeFixture(type)) as CompositeWorkItem;
```
Are we expecting consuming code to be tolerant of null returns? I prefer direct casts when possible because diagnosis at the call site is easier.

CharliePoole replied:
Different philosophy on my part for tests than for production code, I guess. In general, the return is used immediately and we get a null reference exception. I think it's probably six of one etc. unless the return is being held on to for a long time and then used later.

CharliePoole replied:
I added a comment clarifying that the cast is guaranteed - WorkItem created from a TestSuite is always composite.

jnm2 replied:
In my mind, safe cast + "cast is guaranteed" comment can be refactored to direct cast and communicate precisely the same thing. But no big deal.

CharliePoole replied:
Hmm... I don't think it communicates that. IT gives rise to the comment "How do you know this cast won't throw?" There's no satisfying everyone, I guess. 😄

CharliePoole replied:
I changed the return type. Problem solved!

jnm2 replied:
In either case you'd also have to explain _how_ you know it won't throw. But no big deal; I'm satisfied either way. Just exchanging viewpoints!

jnm2 commented on src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs:
```diff
@@ -66,17 +86,17 @@ public ParallelWorkItemDispatcher(int levelOfParallelism)
+            ParallelShift.Assign(new TestWorker(ParallelSTAQueue, "ParallelSTAWorker"));
 
-            var worker = new TestWorker(NonParallelQueue, "Worker#STA_NP");
+            var worker = new TestWorker(NonParallelQueue, "NonParallelWorker");
```
This always confused me! I like the changes.

jnm2 commented on src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs:
```diff
@@ -244,41 +267,45 @@ private void RestoreQueues()
-
-            if (_isolationLevel > 0)
-                RestoreQueues();
+            ShiftFinished?.Invoke(sender as WorkShift);
```
Same thing. Can we direct-cast sender to `WorkShift` or at least throw an `ArgumentException`? Would seem to indicate abuse and safe casts create a secondary breach of expectations (raising ShiftFinished with a null WorkShift) while hiding the root cause.

jnm2 commented on src/NUnitFramework/framework/Internal/Execution/TestWorker.cs:
```diff
@@ -122,10 +122,19 @@ private void TestWorkerThreadProc()
                     Busy(this, _currentWorkItem);
 
+                    // Because we execute the current item AFTER the queue state
+                    // is saved, it's children end up in the new queue set.
```
`it's` should be `its`

jnm2 commented on src/NUnitFramework/framework/Internal/Execution/WorkShift.cs:
```diff
@@ -161,11 +161,11 @@ private void StartWorkers()
                 {
                     // Quick check first using Interlocked.Decrement
-                    if (Interlocked.Decrement(ref _busyCount) == 0)
+                    if (Interlocked.Decrement(ref _busyCount) == 0 && !HasWork)
```
Two concerns with `HasWork` outside a `lock`: 
 - In the worst case scenario, the current thread's view of `HasWork` is stale and never updates to become `false`, so the lock is never taken. We need a memory barrier like `Volatile.Read` or `Interlocked` which brings me to the next point:
 - Reading the `IList` can result in corrupted results (dirty read, tearing- I've observed this in practice) if the `IList` is being changed by another thread. Either we need an absolute guarantee that the `IList` is thread safe (which solves the first point as well), or we need to lock read as well as write access to it (which again solves the first point).

jnm2 commented on src/NUnitFramework/framework/Internal/Execution/WorkShift.cs:
```diff
@@ -161,11 +161,11 @@ private void StartWorkers()
                 {
                     // Quick check first using Interlocked.Decrement
-                    if (Interlocked.Decrement(ref _busyCount) == 0)
+                    if (Interlocked.Decrement(ref _busyCount) == 0 && !HasWork)
```
Two concerns with `HasWork` outside a `lock`: 
 - In the worst case scenario, the current thread's view of the list memory behind `HasWork` is stale and never updates to become `false`, so the `if` branch is never executed. We need a memory barrier which brings me to the next point:
 - Reading the `IList` can result in corrupted results (dirty read, tearing- I've observed the weird internal exceptions caused by this in practice) if the `IList` is being changed by another thread. Either we need an absolute guarantee that the `IList` itself is a thread safe collection (which solves the first point as well), or we need to lock read access as well as write access to it (which again solves the first point).

jnm2 replied:
And I see it's backed by a `List<>` which means we definitely can't read it without locking access.

CharliePoole replied:
This could be a source of some intermittent issues I'm seeing. I'll check.

BTW, this is far from done!

CharliePoole replied:
@jnm2 Thinking about this, I realized that I'm relying on the fact that the underlying list is never changed once set up. A lot of the parallel dispatcher code is like that...

* `ParallelWorkItemDispatcher` has fixed lists of shifts, queues and workers.
* `WorkShift` has a specific set of queues and workers.
* `WorkItemQueue` has a specific set of workers.

For that reason, I've played a little loose with locking. Maybe it would be better to use a collection that can't be updated at all, just in case.

jnm2 replied:
Ah, and I'm so used to looking for `IReadOnlyList<>` to make an appearance for that guarantee... It would simplify my understanding of the code if there was a builder type. How much code would break if `WorkShift` stopped exposing the `IList<>` and made a defensive array copy of the list passed into the constructor, stored as a private readonly `WorkItemQueue[]` and an XML comment on it that it should be immutable? (Or we could even add a lightweight `ImmutableArray` to our BCL-like toolbox?)

So long as no thread mutates the `IList<>` after the first time another thread reads it, we don't need to lock access to it.

CharliePoole replied:
Queues is only public for test purposes. Could easily be internal. I'll take a look at the other collections of queues, shifts and workers as well. This may be an easy fix.

CharliePoole replied:
For now, I'm only changing this to internal. In fact, determining whether a shift is really over is the key to a lot of our issues so I expect to be back in this bit of code in subsequent commits as I add tests for the various issues.

jnm2 replied:
I got sidetracked talking about the list, but we still have a staleness issue checking `WorkItemQueue.IsEmpty`. We need locked access to that mutable state.

jnm2 replied:
[Just checked on it.](https://github.com/nunit/nunit/blob/0162e7c1d8640769be2714ea81a3002903f36617/src/NUnitFramework/framework/Internal/Execution/WorkItemQueue.cs#L169-L179) No staleness issue because you're looking through an array (which I assume is also mot mutated once multiple threads see it) and each array item is a ConcurrentQueue.

Though on a new topic now, having more than one concurrency primitive in action makes me feel like everything is susceptible to race conditions. What happens if one of the queues gets an item added on a background thread right after `IsEmpty` returns false?

jnm2 commented on src/NUnitFramework/tests/Attributes/TestOrderAttributeTests.cs:
```diff
@@ -27,18 +26,15 @@ public void CheckOrderIsCorrect()
+            var testSuite = new TestSuite("dummy").Containing(candidateTypes);
 
-            var work = TestBuilder.PrepareWorkItem(testSuite, null) as CompositeWorkItem;
+            var work = TestBuilder.CreateWorkItem(testSuite) as CompositeWorkItem;
```
Correct me if I'm wrong but my memory is that the return type of `TestBuilder.CreateWorkItem` is already `CompositeWorkItem` s othe `as` is redundant?

CharliePoole replied:
It returns `WorkItem` in this overload. But I'll look at the overloads again.

jnm2 replied:
Taking another look, this does make sense. Though again I think the safe cast clouds intent, whichever way you look at it. But nbd.

jnm2 commented on src/NUnitFramework/tests/Internal/Execution/ParallelExecutionTests.cs:
```diff
@@ -0,0 +1,194 @@
+            var workItem = TestBuilder.CreateWorkItem(_testSuite, context);
+
+            var run_complete = new ManualResetEvent(false);
+            workItem.Completed += (s, e) => { run_complete.Set(); };
```
(`(s, e) => run_complete.Set();` is an option too)

jnm2 replied:
Also `run_complete` should really be in a `using` statement because it holds unmanaged handles.

jnm2 commented on src/NUnitFramework/tests/Internal/Execution/ParallelExecutionTests.cs:
```diff
@@ -0,0 +1,194 @@
+        [Test]
+        public void AllTestsRan()
+        {
+            if(_result.PassCount != _numCases)
```
Inconsistent spacing after the `if` keyword

jnm2 commented on src/NUnitFramework/tests/Internal/Execution/ParallelExecutionTests.cs:
```diff
@@ -0,0 +1,194 @@
+            int count = 0;
+            foreach (var @event in _events)
+            {
+                if (@event.StartsWith("ShiftStarted") && ++count > 1)
```
I guess this is really not important for tests, but if this was shipping code, I'd yell about the missing `StringComparison.Ordinal` parameter because thread culture can wreck string comparisons.

CharliePoole replied:
I agree with both points. 😈

jnm2 commented on src/NUnitFramework/tests/Internal/Execution/ParallelExecutionTests.cs:
```diff
@@ -0,0 +1,194 @@
+                            .Containing(TestBuilder.MakeFixture(typeof(TestSetUpFixture))
+                                .Containing(typeof(TestFixture1), typeof(TestFixture2), typeof(TestFixture3))))),
+                3);
+        }
```
Neat!

jnm2 commented:
Sorry, gotta go but I'll be back.

CharliePoole commented:
I initially worked on this with the idea of completely changing the structure of parallel execution. Test fixtures that needed to run in isolation would use an entirely new dispatcher, the old dispatcher would quiesce until the new one finished and the new dispatcher would get new work shifts, queues and workers.

This turned out to be very complex, requiring intercommunication among the dispatchers. After spending about 20 hours on it, I'm going back to the original design and trying to make it work. That is, `WorkItemQueue` maintains a stack of real queues under the covers. This is used to isolate non-parallel fixtures so that it's impossible for anything else to run at the same time. 

The current / continuing design is a bit ugly in my opinion, but it has been working and I think it's a better use of time to fix the bugs right now.

CharliePoole commented:
Merges are pulling in code that uses the old `TestBuilder` interfaces. I'll rebase again.

jnm2 commented:
GitHub thinks a bunch of my comments are no longer relevant but they are. Should I delete and recomment?

CharliePoole commented:
@jnm2 GitHub does that after a forced push and I just had to do two in a row. Wait till I fix the things you already pointed out - I can see them with a few extra clicks.

jnm2 commented on src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs:
```diff
@@ -244,41 +276,46 @@ private void RestoreQueues()
-            // TODO: Temp fix - revisit later when object model is redesigned.
-            if (StartNextShift())
-                return;
+            ShiftFinished?.Invoke(sender as WorkShift);
```
This one is production code. Can we direct-cast sender to `WorkShift` or at least throw an `ArgumentException`? Would seem to indicate abuse and safe casts create a secondary breach of expectations (raising ShiftFinished with a null WorkShift) while hiding the root cause.

CharliePoole replied:
Changing delegate to be type-safe.

jnm2 commented on src/NUnitFramework/framework/Internal/Execution/WorkItem.cs:
```diff
@@ -248,6 +253,16 @@ public virtual void Execute()
 #endif
         }
 
+        private ManualResetEvent _completionEvent = new ManualResetEvent(false);
```
`ManualResetEvent` is `IDisposable` and therefore the owning class should implement `IDisposable` and dispose it to free the kernel handles.

jnm2 replied:
The field should be `readonly` too.

CharliePoole replied:
Good one!

jnm2 commented on src/NUnitFramework/framework/Internal/Execution/WorkItem.cs:
```diff
@@ -177,6 +177,11 @@ public virtual ParallelExecutionStrategy ExecutionStrategy
+        /// <summary>
+        /// Indicates whether this work item should use a separate dispatcher.
+        /// </summary>
+        public virtual bool IsolateChildTests { get; } = false;
```
We can save an autogenerated field per instance by doing `IsolateChildTests => false;` since we aren't setting `IsolateChildTests` in a constructor.

CharliePoole replied:
I don't love this syntax, especially when the property is virtual. IMO, the existing syntax is clearer to read and avoids confusion.

jnm2 replied:
This is theoretical but C# 7.0 allows `IsolateChildTests { get => false; }` as well, if you still want to use the `get` keyword without allocating a field.

jnm2 commented on src/NUnitFramework/tests/Internal/Execution/ParallelExecutionTests.cs:
```diff
@@ -0,0 +1,196 @@
+        [Test]
+        public void AllTestsRan()
+        {
+            if(_result.PassCount != _numCases)
```
Inconsistent spacing after the `if` keyword

jnm2 commented on src/NUnitFramework/tests/Internal/Execution/ParallelExecutionTests.cs:
```diff
@@ -0,0 +1,196 @@
+            foreach (string @event in _events)
+            {
+                sb.Append(@event);
+                sb.Append(NL);
```
`sb.AppendLine()`?

jnm2 replied:
`new StringBuilder(message).AppendLine();`, and `sb.AppendLine(@event)` here? This would allow you to remove the `NL` declaration too.

jnm2 commented on src/NUnitFramework/tests/Internal/Execution/ParallelExecutionTests.cs:
```diff
@@ -0,0 +1,196 @@
+
+        private string DumpEvents(string message)
+        {
+            var sb = new StringBuilder(message + NL);
```
`new StringBuilder(message).AppendLine();`?

jnm2 commented on src/NUnitFramework/tests/Internal/Execution/ParallelExecutionTests.cs:
```diff
@@ -0,0 +1,196 @@
+    public class ParallelExecutionTests : ITestListener
+    {
+        private TestSuite _testSuite;
+        private int _numCases;
```
Both can be `readonly`.

jnm2 commented:
Pulling the code.

jnm2 commented:
I'm trying to put hte whole picture together but I am feeling uneasy about things like this: Between checking `.HasWork` and returning the shift, another thread may cause `HasWork` to become false.

https://github.com/nunit/nunit/blob/0162e7c1d8640769be2714ea81a3002903f36617/src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs#L312-L318

The next thing that happens is that we call `shift.Start()` on a shift that has potentially just ended:

https://github.com/nunit/nunit/blob/0162e7c1d8640769be2714ea81a3002903f36617/src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs#L299-L304

Which starts a new thread and loops forever until it finds work which may never happen because the work ended between `HasWork` checking it and returning it:

https://github.com/nunit/nunit/blob/0162e7c1d8640769be2714ea81a3002903f36617/src/NUnitFramework/framework/Internal/Execution/WorkItemQueue.cs#L273-L278

Am I missing something or could this cause the thread to busy-wait forever?

jnm2 commented:
All the busy waits I'm seeing could also be contributors to the CPU spikes.

CharliePoole commented:
@jnm2 Regarding HasWork, here's the logic that __should__ apply, unless there is a bug.

* Only one shift ends at a time, so when a shift has ended, no shift is running.
* We only call HasWork when no shift is running
* Since no shift is running, no thread can create more work.

Admittedly, there could be a problem here. However, I'm more suspicious of the correct detection of the end of the shift than I am of the checking logic. If a shift ends while some thread is still running, then all of the assumptions above can be wrong.

What I've done so far is merely creating an infrastructure for us to test all this stuff for real. As I add more test cases now, I'm hoping they will tell us more.

jnm2 commented on src/NUnitFramework/framework/Internal/Execution/WorkItem.cs:
```diff
@@ -300,9 +315,35 @@ public virtual void Cancel(bool force)
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
```
I'm sorry; it crossed my mind if I should mention this originally because MSDN is out of date and misleads people. Real quick: since .NET 4.0 introduced the SafeHandle class, it's now an anti-pattern to use the `Dispose(bool)` pattern or to involve the finalizer. (The SafeHandle contains the finalizer logic and gets other special treatments. Any unmanaged resources (bare IntPtr handles) should go inside a SafeHandle class instead, and the SafeHandle, being itself a managed resource just gets disposed like any other managed object.)

Now I know we run on net35 and net20, but the same logic applies: the `WorkItem` class itself, and any subclass, will not and should not ever directly own bare unmanaged (IntPtr) handles. So as a safeguard we can remove everything except `public void Dispose() { _completionEvent.Dispose(); }`. (And we don't need the null check because the variable is readonly.)

jnm2 replied:
ManualResetEvent on net35 and net20 already contains the finalizer logic too, so it would be redundant for the `WorkItem` class to have finalizer logic. It's already guaranteed to be disposed without `WorkItem` doing anything special.

jnm2 replied:
I know that this is counter to a lot of common practice, so it's understandable that you might want confirmation that I'm not crazy on this. 😀 This comes up a lot at dotnet/csharplang and I can loop in sharwell or someone.

CharliePoole replied:
@jnm2 I won't be changing this right away - it's also an anti-pattern to keep changing minor bits in code that is not yet even doing what it's supposed to do!!! I'll hold it in mind as I work on the next commit with actual functionality.

Based on what you are now saying, it sounds as if the initial code was correct and that WorkItem does not need to implement dispose. I understood that MRE basically took care of itself in different ways on each platform, so we didn't have to. I'm not in favor of implementing an empty Dispose "just in case". Or are you saying we need to implement and invoke DIspose solely for the purpose of getting the release done early?

Please give it some thought and answer at your leisure.

jnm2 replied:
> it's also an anti-pattern to keep changing minor bits in code that is not yet even doing what it's supposed to do!!!

Yes! :D

> Or are you saying we need to implement and invoke DIspose solely for the purpose of getting the release done early?

Yes, exactly. Deterministic release is always preferable to waiting for the garbage collector to finalize an object. One reason is that the garbage collector reacts to memory pressure but not to leaking handles so you can go for an arbitrarily large amount of time without a GC and yet still leaking kernel handles.

The finalization handling is a backstop, not a great one but a necessary one, and since it's built into ManualResetEvent we don't need to build it ourselves. But we'd prefer not to use the backstop.

CharliePoole replied:
Do we believe this to be true on all platforms, including Mono?

jnm2 replied:
I checked Mono and it's using a SafeHandle. (https://github.com/mono/mono/blob/master/mcs/class/referencesource/mscorlib/system/threading/eventwaithandle.cs)

If it's not true in a platform's BCL, it's a bug in the platform just as much as it would be a bug in NUnit if we used a bare IntPtr handle without encapsulating it in a SafeHandle class.

.NET Framework and .NET Core are identical. Mono has been incrementally replacing its BCL with .NET Core's BCL implementation. Xamarin and Unity use Mono's BCL.

CharliePoole replied:
Sounds like the right way to go then. Ironic to be doing what was called bad practice for so long.

jnm2 replied:
This is over 3 years old, but may be relevant: http://www.mono-project.com/docs/advanced/safehandles/

rprouse replied:
@CharliePoole I don't think it is an issue since this is the way we have been doing it, but were you planning on updating this based on the comments?

CharliePoole commented:
This now fixes #2395 

Restoring the queues was starting all queues, causing __everything__ to run in parallel.

CharliePoole commented:
I feel like this is ready for a final review and merge. I imagine there is more to be done around parallel execution, but since this is getting pretty big I think we're better off to put it out there and fix further problems incrementally.

The parallel tests I created can be extended as needed to simulate parallel execution in other environments.

rprouse commented:
I am going through these and writing repros for each of the issues to see if this PR fixes them. So far,

#1239 - I cannot reproduce on my machine, could be number of cores?
#2364 - Unit tests are enough
#2388 - Using the tests in the issue, this is **not fixed**, the given tests still fail with this code
#2395 - This looks to be a duplicate of #2388, so **likely won't be fixed**
#2467 - Confirmed to be a framework problem in 3.8.1 and **fixed** with this PR
#2471 - Confirmed **fixed**
#2478 - I think I've confirmed this by applying `Parallizable` to namespace `OneTimeSetup`s, mixing them with `TestContext.Progress.WriteLine` and `Thread.Sleep` and observing the behavior. Will also look in the code.

So, at this point, assuming the code review finds the same, all issues marked fixed except for #2388 and #2395 are fixed by this PR. We can treat those issues in a separate PR unless you feel they should be included in this PR @CharliePoole.

Code review to follow. I've done a quick pass, now doing a more detailed pass after reading and testing all the issues.

rprouse commented on src/NUnitFramework/framework/Internal/Execution/WorkItemQueue.cs:
```diff
@@ -339,14 +339,60 @@ internal void Save()
         internal void Restore()
         {
-            Pause();
+            //Guard.OperationValid(State != WorkItemQueueState.Running, $"Attempted to restore state of {Name} while queue was running.");
```
Is this a valid guard? Should it be uncommented or deleted?

CharliePoole replied:
It's logically valid (in my mind) but breaks things if I use it. I think I'll keep it commented but add an explanation in front of it. At some point, someone will discover __why__ some queues end up running at a point where we are supposedly between shifts. My guess is that some other thread is getting in their and restarting things.

This is where I'd continue if I had time to do it by the way. We need one thread to detect end of shift and take care of starting next shift, restoring, etc. before any other shift tries to do the same. 

However, I think that overall this fixes a lot of things and is not any worse in the areas it doesn't fix.

rprouse commented on src/NUnitFramework/framework/Internal/Execution/WorkShift.cs:
```diff
@@ -161,16 +170,18 @@ private void StartWorkers()
                         {
-                            // Check busy count again under the lock. If there is no work
-                            // try to restore any saved queues and end the shift.
+                            // Check again under the lock. If there is no work
```
@jnm2 said above,

> having more than one concurrency primitive in action makes me feel like everything is susceptible to race conditions. What happens if one of the queues gets an item added on a background thread right after IsEmpty returns false?

@jnm2 isn't this okay because we check `_busyCount` and `HasWork` again after getting the lock? The comment is pretty obvious, so what you are seeing is more subtle than that. Can you expand on your concern?

CharliePoole replied:
I agree with both of you. I'm worried like @jnm2 and not seeing how it happens like @rprouse. 😄 

My observation is that things get in - it's rare but happens.

CharliePoole commented on src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs:
```diff
@@ -242,49 +278,83 @@ private void RestoreQueues()
-                RestoreQueues();
+            WorkShift nextShift = null;
 
+#if false
```
Slipped through!

CharliePoole replied:
Fixed

CharliePoole commented:
I'll make the changes in the morning. This is far from perfect, but I think it's better than what we have and should be merged so that others can carry on from this point. In particular, the tests can be expanded and possibly the test mechanism improved.

BTW, if #2388 is not fixed, that would be because the simulation (that's what it is) of executing an assembly doesn't fully match the reality. I was hoping to avoid adding true assemblies that have to be loaded when we are working on unit tests. We need a set of tests that run out of actual assemblies to really test what happens with assembly-level attributes but I don't think they should be part of the framework tests.

That said, it may be better to improve the simulation so that we run more of the production code and less `TestBuilder` code. We may be able to inject a context right at the start with a particular (simulated) assembly-level attribute. If nobody else gets too it I may try to take it further in December.

rprouse commented:
I agree that this is an improvement and something we should get in sooner than later. This is also getting huge, so it might be better to split out the remaining issues. I also agree that we need true test assemblies to test some of these things because our unit tests are not catching them. That is why I was suggesting some sort of integration tests that test some of these scenarios. Some of these issues are difficult to even test that way though. For some, I had to add in sleeps and watch the output before and after the fixes.

CharliePoole commented:
I removed the fixes comment for #2388 based on your tests. However, I'm pretty sure that #2395 is fixed. It doesn't seem to be a duplicate of the other at all. According to the extract from the internal trace, multiple threads were running at the same time, when they should not have been. That is, the parallel and non-parallel workers were active simultaneously, which of course defeats their purpose.

I found this was due to an error in how the dispatcher restored queues when we switch from an isolated non-parallel situation back to the lower level parallel one. It was starting all the queues! I fixed this and I think it needs to be checked to see if there is more to fix. Best bet would be to merge and then encourage the reporter of the issue to try the dev build. Since we don't want anyone else to work on it until that is done, it seems best to close it as done.

CharliePoole commented:
@rprouse I made the minor changes you requested. The only open question is whether this should really close #2395. I left that "fixes" in the description, but we can re-open it if it turns out I'm wrong.
